src/modm/platform/core/cortex/delay.cpp.in

```diff
@@ -35,13 +35,28 @@ delay_us(uint32_t us)
 %% else
 	uint32_t start = DWT->CYCCNT;
 	// prefer this for cores with fast hardware multiplication
-	int32_t delay = int32_t(platform::delay_fcpu_MHz) * us - {{ overhead }};
+	int32_t delay = int32_t(platform::delay_fcpu_MHz) * us - {{ overhead * loop }};
 
 	while (int32_t(DWT->CYCCNT - start) < delay)
 		;
 %% endif
 }
 
+void __attribute__((naked, align(4)))
+delay_ns(uint32_t)
+{
+	// ns_per_loop = nanoseconds per cycle times cycles per loop ({{loop}} cycles)
+	asm volatile (
+		".syntax unified"		"\n\t"
+		"lsls.n	r1, %[d], #{{oshift}}"	"\n\t"	// multiply the overhead cycles with the ns per cycle:  1-2 cycles on cm3, up to 32 cycles on cm0
+		"subs.n	r0, r0, r1"		"\n\t"	// subtract the overhead in ns from the input:          1 cycle
+	"1:  subs.n	r0, r0, %[d]"	"\n\t"	// subtract the ns per loop from the input:             1 cycle
+		"bpl.n	1b"				"\n\t"	// keep doing that while result is still positive:      2 cycles (when taken)
+		"bx lr"
+	:: [d] "l" (platform::delay_ns_per_loop) : "r0", "r1");
```

Comment:
I'm pretty sure this inline asm is probably subtly wrong due to missing output variables, I'll check later.